sdks/java/io/solr/src/test/java/org/apache/beam/sdk/io/solr/SolrIOTest.java

```diff
@@ -263,4 +276,109 @@ public void testSplit() throws Exception {
+    AuthorizedSolrClient solrClient = mock(AuthorizedSolrClient.class);
+
+    // simulate success
+    when(solrClient.process(any(String.class), any(SolrRequest.class)))
+        .thenReturn(mock(SolrResponse.class));
+
+    List<SolrInputDocument> batch = SolrIOTestUtils.createDocuments(1);
+    writeFn.flushBatch(solrClient, batch);
+    verify(solrClient, times(1)).process(any(String.class), any(SolrRequest.class));
+  }
+
+  /**
+   * Ensure that the default retrying behavior surfaces errors immediately under failure conditions.
+   */
+  @Test
+  public void testWriteRetryFail() throws Exception {
+    SolrIO.Write write = mock(SolrIO.Write.class);
+    when(write.getRetryConfiguration()).thenReturn(SolrIO.DEFAULT_RETRY_CONFIGURATION);
+    SolrIO.Write.WriteFn writeFn = new SolrIO.Write.WriteFn(write);
+    AuthorizedSolrClient solrClient = mock(AuthorizedSolrClient.class);
+
+    // simulate failure
+    when(solrClient.process(any(String.class), any(SolrRequest.class)))
+        .thenThrow(new SolrServerException("Fail"));
+
+    List<SolrInputDocument> batch = SolrIOTestUtils.createDocuments(1);
+    try {
+      writeFn.flushBatch(solrClient, batch);
+      fail("Error should have been surfaced when flushing batch");
+    } catch (IOException e) {
+      verify(solrClient, times(1)).process(any(String.class), any(SolrRequest.class));
+    }
+  }
+
+  /**
+   * Ensure that a time bounded retrying is observed.
+   */
+  @Test
+  public void testWriteRetryTimeBound() throws Exception {
+    SolrIO.Write write = mock(SolrIO.Write.class);
+    when(write.getRetryConfiguration())
+        .thenReturn(
+            SolrIO.RetryConfiguration.create(Integer.MAX_VALUE, Duration.standardSeconds(3)));
+    SolrIO.Write.WriteFn writeFn = new SolrIO.Write.WriteFn(write);
+    AuthorizedSolrClient solrClient = mock(AuthorizedSolrClient.class);
+
+    // simulate failure
+    when(solrClient.process(any(String.class), any(SolrRequest.class)))
+        .thenThrow(
+            new HttpSolrClient.RemoteSolrException(
+                "localhost", 1, "ignore", new IOException("Network")));
+
+    List<SolrInputDocument> batch = SolrIOTestUtils.createDocuments(1);
+    Stopwatch stopwatch = Stopwatch.createStarted();
+
+    try {
+      writeFn.flushBatch(solrClient, batch);
+      fail("Error should have been surfaced when flushing batch");
+    } catch (IOException e) {
+      // at least two attempts must be made
+      verify(solrClient, Mockito.atLeast(2)).process(any(String.class), any(SolrRequest.class));
+      long seconds = stopwatch.elapsed(TimeUnit.SECONDS);
+      assertTrue(
+          "Retrying should have executed for at least 3 seconds but was " + seconds,
+          seconds >= 3);
+    }
+  }
+
+  /**
+   * Ensure that retries are initiated up to a limited number.
+   */
+  @Test
+  public void testWriteRetryAttemptBound() throws Exception {
+    SolrIO.Write write = mock(SolrIO.Write.class);
```

Comment:
Recent guidelines for IOs encourage to test more at the PTransform level than at the DoFn one which means better use TestPipeline + PAssert et co. Not sure it would be 'easy' to cover these cases like that, but maybe worth a try, at least this one seems straightforward.

Reply:
Thank you very much for the review and guidance @iemejia.  I'll try and accommodate this tomorrow.

Reply:
I'm struggling to find a way to accommodate this.  

The original design was a simple unit test to isolate the `batch()` flushing behaviour and then mock the underlying `SolrClient` simulating failure scenarios. The tests then ensured that for a single batch flush, depending on the configuration, the `SolrClient` was called the correct number of times until retries where exhausted (time bound or N).

Moving this up to a pipeline test introduces 1) an issue with how to simulate failures as I don't (yet) find an easy way to mock the SolrClient (you can't `Mockito.spy()` on the configuration object) and 2) you loose guarantee of the number of times a batch flush or SolrClient should be called as it is in the hands of the runner to present the batches.

Do have any idea I could explore please? I'll keep exploring in the meantime...

I am also not entirely sure why a "mini integration test" using PTransform is better than isolating the methods and testing those through unit tests.  Is it to encourage consistency across tests perhaps?

Reply:
@iemejia - How about I explore moving the retrying behaviour into the `AuthorizedSolrClient` so that it pollutes the SolrIO less and can be tested in isolation of the DoFn and Pipeline?

Reply:
@timrobertson100 I did not see the code, just wanted to comment on your last comment: +1 to moving the retry mecanism.
Also, a general comment about mocks, I'm convinced that mocks are fine to test algorithm but not to fake a middleware otherwise you want have a compelling test. using embeded middlewares for UT is better to be closer from real at a small scale. For larger scale IT are mandatory because some cases like cache or timeouts etc...  only raise at a larger scale especially in big data projects

Reply:
Thanks @echauchot. In this case what I'm trying to test is "does the code that wraps the SOLR client deal with exceptions raised by falling into a retry mechanism". Wouldn't you describe that as an algorithm test?  I'm all for IT as well but in this instance, I'd propose it doesn't call for one as we'd have to either intercept HTTP or have fake / modified servers.

Reply:
@timrobertson100 great idea to move this logic into the `AuthorizedSolrClient` class, way cleaner. About the tests I am still hesitant about the best way to do it because I am also not a big fan of mocks in particular in this case because we have already an embedded version available, but I agree that simulating the unavailability is not trivial.

Reply:
How about the following (copied from the `JdbcIO` design and test approach)

 1. I introduce a configurable `RetryStrategy` that can be added to the `Write` and will come into effect when a the `RetryConfiguration` is also provided
 2. A `DefaultRetryStrategy` will retry on `HttpSolrClient.RemoteSolrException`, `SolrServerException`, `IOException` and also on `SolrException` (where the code is `5xx` only) and I see is missing in the original PR.
 3. I introduce logging using `SLF4J` logging at `warn` when retries are invoked
 4. I run the mini-IT style tests using the embedded Solr server with 
 	1. a custom `RetryStrategy` to return `true`(i.e. retry) on any `SolrException`
 	2. write to a non existing collection which will cause a `SolrException` (code = 400 and a message "collecton not found: ...") thus invoking the retry thanks to the strategy
 	3. Use the same as apprach as [`JdbcIOTest` of checking for log statements](https://github.com/apache/beam/blob/master/sdks/java/io/jdbc/src/test/java/org/apache/beam/sdk/io/jdbc/JdbcIOTest.java#L359) to verify retry behavior is invoked
  
I'd suggest that we keep the 10 lines of retry within the ´SolrIO´ then similar to the ´JdbcIO´ and also meaning it is the `SolrIO` logs we're using in the test.

I'm happy to implement this (especially since it brings further consistency with `JdbcIO`) but would like to have a committer quickly approve the procedure first. Fundamentally we're saying relying on log statements trumps mocking the underlying `SolrClient` (not something I particularly agree with, but keen to follow the project approach).

CC @iemejia @echauchot @jkff

Reply:
@timrobertson100 good idea, I like the fact of being consistent among IOs. Agree in this case also about letting the retry in `SolrIO` to not get into madness with testing on `AutorizedSolrClient'. Please go ahead !

Reply:
And about the second alternative of stopping the server, I am afraid that this introduces some flakiness, so better to have not to do it, It is better to have more reproducible (consistent) tests.

Reply:
Thanks for the quick answer @iemejia - I will do as outlined above.

I'll go for the simple option of initiating retries by writing to a non-existing collection.

My thoughts on the version which would stop the server - It would have to be a separate test class to not interfer with the other tests. I was thinking of starting a monitor thread, running an unbounded stream writing into Solr, having the monitor thread query Solr periodically until it sees data and only then shut it down.  After this, the retry mechanism would kick in and would exhaust retries and then we'd check the logs to verify. While this would be a more real-world test and I think could be done robustly, it is probably unnecessary as the simple option will ensure the test.